torch/csrc/cuda/nccl.cpp

```diff
@@ -0,0 +1,175 @@
+    _communicators[hash.str()] = comms;
+  }
+  return _communicators[hash.str()];
+}
+
+static void _check_inputs(std::vector<at::Tensor> &inputs, std::vector<at::Tensor> &outputs, int size_multiplier) {
+  // len(inputs) == len(outputs)
+  size_t len = inputs.size();
+
+  if (len <= 0) {
+    throw std::runtime_error("input sequence can't be empty");
+  }
+
+  if (len != outputs.size()) {
+    std::stringstream err;
+    err << "inputs and outputs sequences have to be of the same length, but got input of length " << len << " and output of length " << outputs.size();
+    throw std::runtime_error(err.str());
+  }
+
+  std::unordered_set<int> devices;
+  int64_t numel = inputs[0].numel();
+  auto type = inputs[0].type().ID();
+
+  for (size_t i = 0; i < len; i++) {
+    auto input = inputs[i];
+    auto output = outputs[i];
+
+    if (!(input.type().isCuda() && !input.type().isSparse()
+	  && output.type().isCuda()  && !output.type().isSparse())) {
+      throw std::runtime_error("input and output elements have to be cuda Tensors");
+    }
+
+    if (type != input.type().ID() || type != output.type().ID()) {
+      throw std::runtime_error("all inputs and outputs must be of the same Tensor type");
+    }
+
+    if (! (input.is_contiguous() && output.is_contiguous())) {
+      throw std::runtime_error("all inputs and outputs have to be contiguous");
+    }
+
+    // inputs must be on unique devices
+    if (devices.find(input.get_device()) != devices.end()) {
+      throw std::runtime_error("inputs must be on unique devices");
+    }
+    devices.insert(input.get_device());
+
+    // inputs and outputs must be on same device respectively
+    if (input.get_device() != output.get_device()) {
+      throw std::runtime_error("input and output must be on the same device");
+    }
+
+    // all inputs must be same size
+    if (input.numel() != numel) {
+      throw std::runtime_error("all inputs must have the same number of elements");
+    }
+  
+    // outputs have to be of size * size_multiplier
+    if (output.numel() != (numel * size_multiplier)) {
+      throw std::runtime_error("output must be of size input_size * size_multiplier");
+    }
+  }
+}
+
+static ncclDataType_t _get_data_type(at::TypeID type) {
+  switch(type) {
+  case at::TypeID::CUDAFloat   : return ncclFloat;
+  case at::TypeID::CUDAHalf    : return ncclHalf;
+  case at::TypeID::CUDADouble  : return ncclDouble;
+  case at::TypeID::CUDALong    : return ncclInt64;
+  case at::TypeID::CUDAInt     : return ncclInt;
+  case at::TypeID::CUDAChar    : return ncclChar;
+  case at::TypeID::CUDAByte    : return ncclChar;
+  default: throw std::runtime_error("Unconvertible NCCL type");
+  }
+}
+
+PyObject * THCPModule_nccl_reduce(PyObject *self, PyObject *args) {
+  HANDLE_TH_ERRORS
+  PyObject *_inputs, *_outputs, *_streams;
+  int root, op, original_device;
+  int64_t count;
+  ncclDataType_t data_type;
+  ncclComm_t* comm;
+  std::mutex* mutex;
```

Comment:
why do you predeclare everything in here? it's not C89 :D

Reply:
was because of an earlier goto statement, will cleanup